src/controllers/dynakube/version/reconciler.go

```diff
@@ -0,0 +1,121 @@
+	"github.com/spf13/afero"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	TmpCAPath = "/tmp/dynatrace-operator"
+	TmpCAName = "dynatraceCustomCA.crt"
+)
+
+type Reconciler struct {
+	dynakube     *dynatracev1beta1.DynaKube
+	dtClient     dtclient.Client
+	hashFunc     ImageHashFunc
+	timeProvider *kubeobjects.TimeProvider
+
+	fs        afero.Afero
+	apiReader client.Reader
+}
+
+func NewReconciler(dynakube *dynatracev1beta1.DynaKube, apiReader client.Reader, dtClient dtclient.Client, fs afero.Afero, versionProvider ImageHashFunc, timeProvider *kubeobjects.TimeProvider) *Reconciler { //nolint:revive
+	return &Reconciler{
+		dynakube:     dynakube,
+		apiReader:    apiReader,
+		fs:           fs,
+		hashFunc:     versionProvider,
+		timeProvider: timeProvider,
+		dtClient:     dtClient,
+	}
+}
+
+// Reconcile updates the version status used by the dynakube
+func (reconciler *Reconciler) Reconcile(ctx context.Context) error {
+	updaters := []versionStatusUpdater{
+		newActiveGateUpdater(reconciler.dynakube, reconciler.dtClient, reconciler.hashFunc),
+		newOneAgentUpdater(reconciler.dynakube, reconciler.dtClient, reconciler.hashFunc),
+		newCodeModulesUpdater(reconciler.dynakube, reconciler.dtClient),
+	}
+	if reconciler.needsReconcile(updaters) {
+		return reconciler.updateVersionStatuses(ctx, updaters)
+	}
+	return nil
+}
+
+func (reconciler *Reconciler) updateVersionStatuses(ctx context.Context, updaters []versionStatusUpdater) error {
+	dockerConfig, err := reconciler.createDockerConfigWithCustomCAs(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, updater := range updaters {
+		if !reconciler.needsUpdate(updater) {
+			continue
+		}
+		log.Info("updating version status", "updater", updater.Name())
+		err := reconciler.run(ctx, updater, dockerConfig)
+		if err != nil {
+			log.Error(err, "failed to update version status", "updater", updater.Name())
+		}
+	}
+	return nil
+}
+
+func (reconciler *Reconciler) createDockerConfigWithCustomCAs(ctx context.Context) (*dockerconfig.DockerConfig, error) {
+	caCertPath := path.Join(TmpCAPath, TmpCAName)
+	dockerConfig := dockerconfig.NewDockerConfig(reconciler.apiReader, *reconciler.dynakube)
+	err := dockerConfig.SetupAuths(ctx)
+	if err != nil {
+		log.Info("failed to set up auths for image version checks")
+		return nil, err
+	}
+	if reconciler.dynakube.Spec.TrustedCAs != "" {
+		_ = os.MkdirAll(TmpCAPath, 0755)
+		err := dockerConfig.SaveCustomCAs(ctx, reconciler.fs, caCertPath)
+		if err != nil {
+			log.Info("failed to save CAs locally for image version checks")
+			return nil, err
+		}
+		defer func() {
+			_ = os.Remove(TmpCAPath)
+		}()
+	}
+	return dockerConfig, nil
+}
+
+func (reconciler *Reconciler) needsReconcile(updaters []versionStatusUpdater) bool {
+	for _, updater := range updaters {
+		if reconciler.needsUpdate(updater) {
+			return true
+		}
+	}
+	return false
+}
+
+func (reconciler *Reconciler) needsUpdate(updater versionStatusUpdater) bool {
```

Comment:
should be moved to `updater` since only the `timeProvider` is passed in from the `reconciler`